Feature/list subdomains

jcnelson opened the pull request:
This patch adds a `GET /list/:pageNum` which lets a caller fetch up to 100 previously-enqueued subdomain requests per `pageNum` (starting with `/list/0`).  Subdomains will be returned in descending order by arrival time, so you get the most-recent subdomains first.  This PR includes unit tests for the `server.js` entry point for listing subdomains.  In addition, this PR can be tested in the `develop` branch of Blockstack Core by running the `subdomain_registrar_list_subdomains` integration test.

This patch also fixes a regression in the unit tests where newer versions of blockstack.js and bitcoinjs-lib don't like the mocked UTXOs.  The fix is simple: use distinct owner and payment keys with distinct input txids.

kantai commented:
This worked great -- I pushed a fix for the error handling cases, the code before would return before writing the HTTP response when you supplied a non-number page (e.g., `/list/abc` would just hang), and the change to the subdomainInfo response would cause the server to write to the socket after it closed.

One question: did we want to limit the response set to just recnt-ish names? We could filter the result set by `received_ts` and use a large default like 1-week.

jcnelson commented:
Awesome, thanks for reviewing!

> One question: did we want to limit the response set to just recent-ish names? We could filter the result set by received_ts and use a large default like 1-week.

Actually, it might make more sense to select records by a UNIX epoch time instead of a page number.  SQL queries like `SELECT * FROM ... OFFSET ? LIMIT ?` take O(n) time in the number of records, whereas `SELECT * FROM ... WHERE cond` takes sublinear time.

Would that be okay with you?  Implementing `GET /list/:unixTime` to return the last 100 rows from the given date instead of implementing `GET /list/:pageNum`?

jcnelson commented:
Okay, I went ahead and implemented `GET /list/:unixTime` instead of `GET /list/:page`.  I also updated the Blockstack Core test in `develop` to reflect this.

kantai commented:
Updated to use the queue-index instead of the  unixTime -- this is a little simpler of an interface. I also added an index on the `received_ts`, as the query now has to do a select based on that. Thanks for getting this PR together @jcnelson.

Merging to develop!
